src/fpump.jl

```diff
@@ -0,0 +1,312 @@
+    mip_model = Model(solver=m.mip_solver)
+    lb = [m.l_var; -1e6]
+    ub = [m.u_var; 1e6]
+    @variable(mip_model, lb[i] <= mx[i=1:m.num_var] <= ub[i])
+    for i=1:m.num_var
+        if m.var_type[i] == :Int
+            setcategory(mx[i], :Int)
+        elseif m.var_type[i] == :Bin
+            setcategory(mx[i], :Bin)
+        end
+    end
+    
+    for c in aff
+        if c.sense == :(>=)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) >= c.rhs)
+        elseif c.sense == :(<=)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) <= c.rhs)
+        elseif c.sense == :(==)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) == c.rhs)
+        end
+    end
+
+    @variable(mip_model, mabsx[i=1:m.num_int_bin_var] >= 0)
+    for i=1:m.num_int_bin_var
+        vi = m.int2var_idx[i]
+        @constraint(mip_model, mabsx[i] >= mx[vi]-nlp_sol[vi])
+        @constraint(mip_model, mabsx[i] >= -mx[vi]+nlp_sol[vi])
+    end
+      
+    # How long is the tabu list 
+    num_sols = 0
+    for i=1:tabu_list.length
+        if !isnan(tabu_list.sols[i][1]) 
+            num_sols += 1
+        else 
+            break
+        end
+    end
+
+    # If there solutions in the tabu list => avoid them
+    if num_sols > 0
+        M = 100
+        @variable(mip_model, z1[j=1:m.num_int_bin_var,k=1:num_sols], Bin)
+        @variable(mip_model, z2[j=1:m.num_int_bin_var,k=1:num_sols], Bin)
+        v = tabu_list.sols
+        for k=1:num_sols, j=1:m.num_int_bin_var
+            i = m.int2var_idx[j] 
+            @constraint(mip_model, z1[j,k]+z2[j,k] <= 1)
+            @constraint(mip_model, (m.l_var[i] - v[k][j])*z1[j,k]+z2[j,k]+v[k][j] <= mx[i])
+            @constraint(mip_model, mx[i] <= v[k][j] - z1[j,k] + (m.u_var[i]-v[k][j])*z2[j,k])
+        end
+        for k=1:num_sols
+            @constraint(mip_model, sum(z1[j,k]+z2[j,k] for j=1:m.num_int_bin_var) >= 1)
+        end
+    end
+
+    @objective(mip_model, Min, sum(mabsx[i] for i=1:m.num_int_bin_var))
+    
+    # Break the mip solver if it takes too long or throw a warning when this option isn't available
+    try 
+        MathProgBase.setparameters!(m.mip_solver, TimeLimit=m.options.feasibility_pump_time_limit)
+    catch
+        warn("Set parameters is not supported")
+    end
+    
+    status = solve(mip_model)
+    println("MIP Status: ", status)
+    println("MIP Obj: ", getobjectivevalue(mip_model))
+
+    # round mip values
+    values = getvalue(mx)
+    for i=1:m.num_int_bin_var
+        vi = m.int2var_idx[i]
+        values[vi] = round(values[vi])
+    end
+    return status, values
+end
+
+"""
+    generate_nlp(m, mip_sol)
+
+Generates the original nlp but changes the objective to minimize the distance to the mip solution
+"""
+function generate_nlp(m, mip_sol)
+    nlp_model = Model(solver=m.nl_solver)
+    lb = [m.l_var; -1e6]
+    ub = [m.u_var; 1e6]
```

Comment:
why are there extra bounds at the end of this array?

Reply:
and below as well?

Reply:
Left over from my early beginning with julia and copying from Katana or something. Thanks for pointing out ;)